torch/distributed/distributed_c10d.py

```diff
@@ -2139,15 +2139,35 @@ def all_gather_coalesced(
     ]
 
     if group is None:
-        default_pg = _get_default_group()
-        work = default_pg.allgather_coalesced(output_tensor_lists, input_tensor_list)
-    else:
-        work = group.allgather_coalesced(output_tensor_lists, input_tensor_list)
-
+        group = _get_default_group()
+    
+    #work = group.allgather_coalesced(output_tensor_lists, input_tensor_list)
+    num_proc = len(output_tensor_lists)
+    
+    fut_list = []
+    temp_out = [[] for _ in range(len(input_tensor_list))]
+    
+    def copy_values_to_output_tensor_lists(fut_param):
+        vals = fut_param.value()
+        print(f"first:{vals} - {fut_param.done()}")
+        print(f"second: {vals[0].value()}")
+    
+    for idx, tensor in enumerate(input_tensor_list):
+        temp_out_list = [torch.zeros(tensor.shape) for _ in range(num_proc)]
+        temp_out[idx] = temp_out_list
+        work = group.allgather([temp_out[idx]], [tensor])
```

Comment:
IIUC, you need to keep the temporary list alive, as the pybind and C++ side is only getting a reference of this list:

https://github.com/pytorch/pytorch/blob/64d3c7388f3ebb084c6e6998597e94753fde1ffa/torch/csrc/distributed/c10d/init.cpp#L1062-L1071

https://github.com/pytorch/pytorch/blob/64d3c7388f3ebb084c6e6998597e94753fde1ffa/torch/csrc/distributed/c10d/ProcessGroupGloo.hpp#L246-L249

Reply:
NVM, I was wrong. I saw the work object is making copies of the list:

https://github.com/pytorch/pytorch/blob/64d3c7388f3ebb084c6e6998597e94753fde1ffa/torch/csrc/distributed/c10d/ProcessGroupGloo.cpp#L1712-L1713